src/librepgp/stream-parse.c

```diff
@@ -495,12 +581,554 @@ encrypted_src_close(pgp_source_t *src)
         free(src->param);
         src->param = NULL;
     }
-    if (src->cache) {
-        free(src->cache);
-        src->cache = NULL;
+}
+
+static pgp_hash_t *
+signed_get_hash(pgp_source_signed_param_t *param, pgp_hash_alg_t alg)
+{
+    for (list_item *hash = list_front(param->hashes); hash; hash = list_next(hash)) {
+        if (pgp_hash_alg_type((pgp_hash_t *) hash) == alg) {
+            return (pgp_hash_t *) hash;
+        }
+    }
+
+    return NULL;
+}
+
+static void
+signed_add_hash(pgp_source_signed_param_t *param, pgp_hash_alg_t alg)
+{
+    pgp_hash_t hash = {0};
+
+    if (!signed_get_hash(param, alg)) {
+        if (!pgp_hash_create(&hash, alg)) {
+            RNP_LOG("failed to initialize hash algorithm %d", (int) alg);
+        } else if (!list_append(&param->hashes, &hash, sizeof(hash))) {
+            RNP_LOG("allocation failed");
+        }
     }
 }
 
+static bool
+signed_validate_signature(pgp_source_t *src, pgp_signature_t *sig, pgp_pubkey_t *key)
+{
+    pgp_hash_t *               hash;
+    pgp_hash_t                 shash = {0};
+    uint8_t                    trailer[6];
+    uint8_t                    hval[PGP_MAX_HASH_SIZE];
+    unsigned                   len;
+    pgp_source_signed_param_t *param = src->param;
+    bool                       ret = false;
+
+    /* Get the hash context */
+    if ((hash = signed_get_hash(param, sig->halg)) == NULL) {
+        RNP_LOG("hash context %d not found", (int) sig->halg);
+        return false;
+    }
+
+    if (!pgp_hash_copy(&shash, hash)) {
+        RNP_LOG("failed to clone hash context");
+        return false;
+    }
+
+    /* hash signature fields and trailer */
+    pgp_hash_add(&shash, sig->hashed_data, sig->hashed_len);
+
+    if (sig->version > 3) {
+        trailer[0] = sig->version;
+        trailer[1] = 0xff;
+        STORE32BE(&trailer[2], sig->hashed_len);
+        pgp_hash_add(&shash, trailer, 6);
```

Comment:
It's memory leak here, isn't? pgp_hash_create gets called once again

Reply:
Errm, where do you see pgp_hash_create? `pgp_hash_add` is a define for `botan_hash_update`

Reply:
Sorry, misunderstanding. I've mixed this function with ``signed_add_hash``